kernel/arch/dreamcast/sound/snd_iface.c

```diff
@@ -66,57 +61,53 @@ int snd_init(void) {
 void snd_shutdown(void) {
     if(initted) {
         spu_disable();
-        sem_destroy(&sem_qram);
         snd_mem_shutdown();
         initted = 0;
     }
 }
 
 /* Submit a request to the SH4->AICA queue; size is in uint32's */
 int snd_sh4_to_aica(void *packet, uint32_t size) {
-    uint32_t  qa, bot, start, top, *pkt32, cnt;
-    assert_msg(size < 256, "SH4->AICA packets may not be >256 uint32's long");
+    uint32_t qa, bot, start, top, *pkt32, cnt;
+    assert_msg(size < AICA_CMD_MAX_SIZE, "SH4->AICA packets may not be >256 uint32's long");
 
-    sem_wait(&sem_qram);
+    g2_ctx_t ctx = g2_lock();
 
     /* Set these up for reference */
     qa = SPU_RAM_UNCACHED_BASE + AICA_MEM_CMD_QUEUE;
-    assert_msg(g2_read_32(qa + offsetof(aica_queue_t, valid)), "Queue is not yet valid");
+    assert_msg(g2_read_32_raw(qa + offsetof(aica_queue_t, valid)), "Queue is not yet valid");
```

Comment:
I am not sure that those assert_msg() are safe, since you are inside a critical section now.

Reply:
If the assert initiated, it breaks the entire program anyway.

Reply:
Fair enough.